src/Build/Graph/ProjectGraph.cs

```diff
@@ -75,6 +76,16 @@ public sealed class ProjectGraph
 
         public IReadOnlyCollection<ProjectGraphNode> GraphRoots { get; }
 
+        /// <summary>
+        ///     Gets a distinct enumeration of all of the environment variables reads during evaluation.
+        /// </summary>
+        public IEnumerable<string> EnvironmentVariableReads => _projectNodesTopologicallySorted?.Value.Select(p => p.ProjectInstance).SelectMany(pi => pi.EnvironmentVariableReads).Distinct();
```

Comment:
These will force the lazy to evaluate eagerly, essentially cancelling the benefit of the lazy. Maybe make these lazy too, which would also avoid recomputing them on each access. Or use `ProjectNodes` which is the same information, but not sorted.

Reply:
Updated to use ProjectNodes.